File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    num_friends = len(friends)
    name_idx_map = {name: idx for idx, name in enumerate(friends)}

    # 1. 선물 주고받은 횟수 기록
    gift_count = [[0] * num_friends for _ in range(num_friends)]
    sent_count = [0] * num_friends
    received_count = [0] * num_friends

    for rec in gifts:
        sender, receiver = rec.split()
        sender_idx = name_idx_map[sender]
        receiver_idx = name_idx_map[receiver]

        gift_count[sender_idx][receiver_idx] += 1
        sent_count[sender_idx] += 1
        received_count[receiver_idx] += 1

    # 2. 선물 지수: (보낸 횟수) – (받은 횟수)
    gift_score = [sent_count[i] - received_count[i] for i in range(num_friends)]

    # 3. 다음 달에 받을 예상 선물 수
    next_month_receive = [0] * num_friends
    for i in range(num_friends):
        for j in range(i + 1, num_friends):
            sent_i_to_j = gift_count[i][j]
            sent_j_to_i = gift_count[j][i]

            if sent_i_to_j > sent_j_to_i:
                next_month_receive[i] += 1
            elif sent_i_to_j < sent_j_to_i:
                next_month_receive[j] += 1
            else:
                # 동률인 경우, gift_score 비교
                if gift_score[i] > gift_score[j]:
                    next_month_receive[i] += 1
                elif gift_score[i] < gift_score[j]:
                    next_month_receive[j] += 1
                # 같으면 증가 없음

    return max(next_month_receive)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
    public int solution(String[] friends, String[] gifts) {
        // 1. 초기화: 이름을 키로 하는 Map 생성
        Map<String, Integer> sentCount = new HashMap<>();
        Map<String, Integer> receivedCount = new HashMap<>();
        Map<String, Map<String, Integer>> giftCount = new HashMap<>();

        for (String name : friends) {
            sentCount.put(name, 0);
            receivedCount.put(name, 0);
            giftCount.put(name, new HashMap<>());
            for (String other : friends) {
                giftCount.get(name).put(other, 0);
            }
        }

        // 2. 선물 주고받은 횟수 기록
        for (String rec : gifts) {
            String[] parts = rec.split(" ");
            String sender = parts[0];
            String receiver = parts[1];

            sentCount.put(sender, sentCount.get(sender) + 1);
            receivedCount.put(receiver, receivedCount.get(receiver) + 1);
            giftCount.get(sender).put(receiver, giftCount.get(sender).get(receiver) + 1);
        }

        // 3. 선물 지수 계산: 보낸 횟수 – 받은 횟수
        Map<String, Integer> giftScore = new HashMap<>();
        for (String name : friends) {
            giftScore.put(name, sentCount.get(name) - receivedCount.get(name));
        }

        // 4. 다음 달에 받을 예상 선물 수
        Map<String, Integer> nextReceive = new HashMap<>();
        for (String name : friends) {
            nextReceive.put(name, 0);
        }

        int n = friends.length;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                String a = friends[i];
                String b = friends[j];

                int a2b = giftCount.get(a).get(b);
                int b2a = giftCount.get(b).get(a);

                if (a2b > b2a) {
                    nextReceive.put(a, nextReceive.get(a) + 1);
                } else if (a2b < b2a) {
                    nextReceive.put(b, nextReceive.get(b) + 1);
                } else {
                    int scoreA = giftScore.get(a);
                    int scoreB = giftScore.get(b);
                    if (scoreA > scoreB) {
                        nextReceive.put(a, nextReceive.get(a) + 1);
                    } else if (scoreA < scoreB) {
                        nextReceive.put(b, nextReceive.get(b) + 1);
                    }
                }
            }
        }

        // 5. 최대값 반환
        return nextReceive.values()
            .stream()
            .max(Integer::compareTo)
            .orElse(0);
    }
}